File: OlderCodes/Fortran77codes/integ.f
c  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
c  by RH Landau, MJ Paez, and CC BORDEIANU 
c  Copyright Princeton University Press, Princeton, 2008.
c  Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
c  MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
c  Support by National Science Foundation                               
c    			 
c  integ.f: Integrate exp(-x) using trapezoid, Simpson & Gauss rules
c           gauss.f must be included 
c           derivation from the exact output as x y1 y2           

      Program integrate
			
      	Implicit none
      	Real*8 trapez, simpson, quad, r1, r2, r3
      	Real*8 theo, vmin, vmax
      	Integer i
c                               theoretical result, integration range
      	theo = 0.632120558829
      	vmin = 0.
      	vmax = 1.
      	Open(6, File = 'integ.dat', Status = 'Unknown')
c                 calc integral using both methods for steps = 3..501
        Do 50 i = 3, 501 , 2
          r1 = trapez(i, vmin, vmax)
          r1 = abs(r1-theo)
          r2 = simpson(i, vmin, vmax)
          r2 = abs(r2-theo)
          r3 = quad(i, vmin, vmax)
          r3 = abs(r3-theo)
          write(6, *) i, r1, r2, r3      
 50     Continue
 	      Close(6)
        Stop 'data saved in integ.dat'
      End
c                                               function to integrate
      Function f(x)
        Implicit none
      	Real*8 f, x
        f = exp(-x)
       	Return
      End
c                                                      trapezoid rule
      Function trapez(i, min, max)
      	Implicit none
      	Integer i, n				
      	Real*8 f, interval, min, max, trapez, x
      	trapez = 0		 
      	interval = ((max-min) / (i-1))
c                                                       sum midpoints
      	Do 21 n = 2, (i-1)          	
          x = interval * (n-1)
          trapez = trapez + f(x)*interval
 21   	Continue 
c                                                       add endpoints
      	trapez = trapez+0.5*(f(min)+f(max))*interval
        Return
      End
c                                                      Simpsons rule
      Function simpson(i, min, max)
      	Implicit none
      	Integer i, n				
      	Real*8 f, interval, min, max, simpson, x
      	simpson = 0		 
      	interval = ((max-min) / (i-1))
c                                                 loop for odd points
      	Do 31 n = 2, (i-1), 2          	
          x = interval * (n-1)
          simpson = simpson + 4*f(x)
 31   	Continue 
c                                                loop for even points
      	Do 32 n = 3, (i-1), 2          	
          x = interval * (n-1)
          simpson = simpson + 2*f(x)
 32   	Continue  
c                                                       add endpoints
      	simpson = simpson+f(min)+f(max)
      	simpson = simpson*interval/3
        Return
      End
c                                                    Gauss quadrature
      Function quad(i, min, max)
      	Implicit none
      	Real*8 w(1000), x(1000)
      	Real*8 f, min, max, quad
      	Integer i, job, n
      	quad = 0
      	job = 0
      	call gauss(i, job, min, max, x, w)
      	Do 41 n = 1, i
          quad = quad+f(x(n))*w(n)
 41   	Continue
      	Return 
      	End

File: OlderCodes/Ccodes/integ.c
/* From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
   Support by National Science Foundation                              
   */
//   integ.c:  Integration using trapezoid, Simpson and Gauss rules
/* comment: The derivation from theoretical result for each method  
            is saved in x y1 y2 y3 format.            
            Program needs gauss.c in the same directory.         */  
#include <stdio.h>
#include <math.h>
#include "gauss.c"                   // returns Legendre pts, weights
#define max_in  501                        // max number of intervals
#define vmin 0.0                             // ranges of integration
#define vmax 1.0      
#define ME 2.7182818284590452354E0                  // Euler's number

main()  {
	
  int i;      
  float result;
  float f(float x);               
  float trapez  (int no, float min, float max);     // trapezoid rule
  float simpson (int no, float min, float max);     // Simpson's rule
  float gaussint(int no, float min, float max);        // Gauss' rule
  
  FILE *output;                             // save data in integ.dat
  output= fopen("integ.dat","w");
	                                          // Simpson requires odd N
  for (i = 3; i<= max_in; i+= 2) {                
    result= trapez(i, vmin, vmax);
    fprintf(output, "%i\t%e\t", i, fabs(result-1+1/ME));
    result= simpson(i, vmin, vmax);
    fprintf(output, "%e\t", fabs(result-1+1/ME));
    result= gaussint(i, vmin, vmax);
    fprintf(output, "%e\n", fabs(result-1+1/ME));
  }
  printf("data stored in integ.dat\n");
  fclose(output);
}                                                              // end  

float f (float x)                            // function to integrate
{ return (exp(-x));  }

float trapez (int no, float min, float max) {            
	                                                 // trapezoid rule
  int n;       
  float interval, sum = 0., x;    
  interval= ((max-min) / (no-1));
  for (n = 2; n<no; n++)   {                     // sum the midpoints
    x   = interval * (n-1);      
    sum += f(x)*interval;
  }
  sum += 0.5 *(f(min) + f(max)) * interval;      // add the endpoints
  return (sum);
}      

float simpson (int no, float min, float max) {           
	                                                  // Simpson's rule
  int n;        
  float interval, sum = 0., x;
  interval= ((max -min) /(no-1));
  for (n = 2; n<no; n+= 2) {                   // loop for odd points
    x= interval * (n-1);
    sum += 4 * f(x);
  }
  for (n = 3; n<no; n+= 2)   {                // loop for even points
    x= interval * (n-1);
    sum += 2 * f(x);
  }   
  sum +=  f(min) + f(max);                // add first and last value
  sum *= interval/3.;             
  return (sum);
}  


float gaussint (int no, float min, float max) {        // Gauss' rule
  int n;        
  float quadra= 0.;
  double w[1000], x[1000]; 
  void gauss(int npts, int job, double a, double b, double x[], double w[]);                       // for points and weights
  gauss (no, 0, min, max, x, w);       // Gauss Legendre points & wts
  for (n = 0; n< no; n++) quadra += f(x[n])*w[n];    // calc integral
  return (quadra);                  
}
